### DISCUSSION TITLE
How to get Terminal instance from a URI

### DISCUSSION THREAD

**Discussion by mo10** [microsoft/vscode-discussions#1065] [Extension Development QnA] [answered]
Hi, I'm currently  coding extension for vscode terminal. I want to add some item to the terminal menu and then send some thing to the terminal.

I can now add a new menu item for terminal instances of the editor tab and terminal panel.

package.json
```json
"menus": {
  "terminal/title/context": [
    {
      "command": "myextension.execute"
    }
  ],
  "editor/title/context": [
    {
      "command": "myextension.execute",
      "when": "resourceScheme == 'vscode-terminal'",
      "group": "navigation"
    }
  ]
}
```
![image](https://github.com/microsoft/vscode-discussions/assets/9402279/977e2089-2c61-434c-a1f4-d4c2b2b1a50c)
![image](https://github.com/microsoft/vscode-discussions/assets/9402279/d911546c-0b80-4cc9-b5e0-3628ba926417)

I can receive `vscode.Terminal` instance from the terminal panel menu. But I only receive `URI` or `vscode.URi` from editor tab.

Example code:
```ts
context.subscriptions.push(vscode.commands.registerCommand('myextension.execute', (args)=>{
    if (args instanceof vscode.Uri) {
        console.log(`is vscode.Uri: ${args}`); // is vscode.Uri: vscode-terminal:/ext-dev/2
    } else {
        let terminal: vscode.Terminal = args;
        terminal.sendText('Hello', false);
    }
}));
```
![image](https://github.com/microsoft/vscode-discussions/assets/9402279/1b49ca3e-1321-4b18-8e39-68469aade47b)

How can I get Terminal instance from this uri? I don't want to place many icons in "editor/title" or "view/title".

Thanks.

**Comment by isidorn** [comment]
Maybe @tyriar might be able to help

**Comment by Tyriar** [comment, marked as answer]
I don't think this is possible currently, the URI was added purely to support opening terminals in the editor area. Adding uri to the Terminal object would be new feature work

**Reply by mo10** [reply]
Thanks @isidorn   @Tyriar

I see that the code for 'Move Terminal into Panel' seems to be able to retrieve the terminal instance through the URI:
https://github.com/microsoft/vscode/blob/b9110d3ad13e4f64da02fb7184747e8768a3de4a/src/vs/workbench/contrib/terminal/browser/terminalEditorService.ts#L202
Maybe need some way to export this.

**Reply by Tyriar** [reply]
@mo10 it would need to go through our API stabilization process which starts with a feature request if you want to make that: https://github.com/microsoft/vscode/issues/new
